### PR TITLE
added to run subprocess in the background in ssh.py

### PR DESCRIPTION
**Pull request by smruti77**
added for run subprocess in the background in remote machine.

Signed-off-by: bismurti bidhibrata pattajoshi <bbidhibr@in.ibm.com>

### CODE REVIEW

File: avocado/utils/ssh.py [hunk 1]
```diff
@@ -155,11 +156,17 @@ def cmd(self, command):
...
         :param command: the command to execute over the SSH session
         :param command: str
+        :param command: when True run subprocess in the background
+        :param bg_processes: bool
         :returns: The command result object.
         :rtype: A :class:`CmdResult` instance.
         """
         cmd = self._ssh_cmd(self.DEFAULT_OPTIONS, ('-q', ), command)
-        return process.run(cmd, ignore_status=True)
+        sub_process = SubProcess(cmd)
```
Comment (PraveenPenguin) [inline avocado/utils/ssh.py]: this line should also be moved in if condition as it only makes sense when user want to run in command  background but the only tricky part is how we terminate the command

File: avocado/utils/ssh.py [hunk 2]
```diff
@@ -155,11 +156,17 @@ def cmd(self, command):
 
         :param command: the command to execute over the SSH session
         :param command: str
+        :param command: when True run subprocess in the background
```
Comment (willianrampazzo) [inline avocado/utils/ssh.py]: I think this line is related to the `bg_processes` and not to `command` parameter. Also, I suggest you use a more meaningful parameter like `background` or `run_background`.

File: avocado/utils/ssh.py [hunk 3]
```diff
@@ -155,11 +156,17 @@ def cmd(self, command):
 
         :param command: the command to execute over the SSH session
         :param command: str
```
Comment (willianrampazzo) [inline avocado/utils/ssh.py]: If possible, also change this documentation as it should be `:type command: str`.

File: avocado/utils/ssh.py [hunk 4]
```diff
@@ -155,11 +156,17 @@ def cmd(self, command):
...
+        :param command: when True run subprocess in the background
+        :param bg_processes: bool
```
Comment (willianrampazzo) [inline avocado/utils/ssh.py]: Here it should also be `:type bg_processes: bool`, if you keep the same parameter name.

File: avocado/utils/ssh.py (hunk 1 above)
Comment (willianrampazzo) [reply avocado/utils/ssh.py]: As the `process` module is already imported, `sub_process = process.SubProcess(cmd)` makes the code more consistent and avoid the need to do line 10 `from . process import SubProcess`.

Comment (smruti77) [reply avocado/utils/ssh.py]: ok, i will chnage that.

**Comment by smruti77** [comment]
> One big problem with this code is that it changes the return value, depending on the newly introduced parameter. That's because `SubProcess.start()` will return a PID, and not a `CmdResult`:
> 
> https://avocado-framework.readthedocs.io/en/75.1/api/utils/avocado.utils.html#avocado.utils.process.SubProcess.start
> 
> @smruti77 it would be nice if you could provide an example of what you're trying to accomplish. I'm specially curious if you're trying to not block on the remote side or on the local side.
> 
> Thanks!

@clebergnu actually we are getting problem with background command, like in iperf, uperf while we are starting the server in peer machine in that command it stucking their and not coming out from that command unable to complete the script.because when we start the server in peer machine it will not be proceed till we run the command on host.we are facing this problem in both the test so now tried with subProcess() its woking fine.

**Comment by pevogam** [comment]
I just want to refer some of you here to a comment I did in https://github.com/avocado-framework/avocado/commit/7475725e1ec5ae089d4ef4d54ba7c8b7a5664614#commitcomment-37399334 and hope that you agree it is best to keep development on this feature to the actual minimum and if possible add a docstring clarifying the same that will at least point some developers in the right feature request direction.

**Comment by clebergnu** [comment]
Hi @smruti77 , I have a counter proposal, considering some of the points that I raised (different return types for `run()`) and most importantly, the observations made by @pevogam .

Please take a look here: https://github.com/clebergnu/avocado/commit/7c6cca4e02e8994fe70a5ba871587d190bae980d

It exposes the command line generation, which is basically the value of this module, and with that you can do whatever you want (like starting multiple processes with the same or different commands).  I've written an example script too.

Please let me know what you all think.

**Comment by smruti77** [comment]
> Hi @smruti77 , I have a counter proposal, considering some of the points that I raised (different return types for `cmd()`) and most importantly, the observations made by @pevogam .
> 
> Please take a look here: [clebergnu@7c6cca4](https://github.com/clebergnu/avocado/commit/7c6cca4e02e8994fe70a5ba871587d190bae980d)
> 
> It exposes the command line generation, which is basically the value of this module, and with that you can do whatever you want (like starting multiple processes with the same or different commands). I've written an example script too.
> 
> Please let me know what you all think.

@clebergnu Hi i checked that PR that is working fine for me.if you want to continue with that approach its fine.

**Comment by smruti77** [comment]
> @smruti77 still I am worrying about how we handling background process we started but that should be killed, even what @clebergnu proposed that seems me a good idea

@PraveenPenguin in my case no need to kill after completed of host command its kill automatic so i didn't added anything for stop.

**Comment by PraveenPenguin** [comment]
> > @smruti77 still I am worrying about how we handling background process we started but that should be killed, even what @clebergnu proposed that seems me a good idea
> 
> @PraveenPenguin in my case no need to kill after completed of host command its kill automatic so i didn't added anything for stop.

@smruti77 sure, But if we adding something in the library we need to handle for all the use cases :)

**Comment by smruti77** [comment]
> > > @smruti77 still I am worrying about how we handling background process we started but that should be killed, even what @clebergnu proposed that seems me a good idea
> > 
> > 
> > @PraveenPenguin in my case no need to kill after completed of host command its kill automatic so i didn't added anything for stop.
> 
> @smruti77 sure, But if we adding something in the library we need to handle for all the use cases :)

@PraveenPenguin ok

**Comment by clebergnu** [comment]
> > > @smruti77 still I am worrying about how we handling background process we started but that should be killed, even what @clebergnu proposed that seems me a good idea
> > 
> > 
> > @PraveenPenguin in my case no need to kill after completed of host command its kill automatic so i didn't added anything for stop.
> 
> @smruti77 sure, But if we adding something in the library we need to handle for all the use cases :)

Since you can create a `SubProcess` instance, like in the example included, methods such as:

https://avocado-framework.readthedocs.io/en/76.0/api/utils/avocado.utils.html#avocado.utils.process.SubProcess.kill

and

https://avocado-framework.readthedocs.io/en/76.0/api/utils/avocado.utils.html#avocado.utils.process.SubProcess.terminate

Can be used without any complexity.

**Comment by clebergnu** [comment]
So, given that the original contributor @smruti77 is fine with the counter proposal, I'm closing this one and moving the discussion to #3602 . Thanks for the feedback!

**Pull request closed without merging**